File: ONE_DIM_OPTIMIZATION/goldenratio.py
import numpy as np
from math import sqrt
import matplotlib.pyplot as plt


def f(x: float) -> float:
    return x ** 2 + 3 * x * (np.log(x) - 1)


def golden_ratio(a: float, b: float, eps: float) -> tuple:
    k: int = 1
    phi: float = (1 + sqrt(5)) / 2
    x1: float = b - (b - a) / phi
    x2: float = a + (b - a) / phi
    while abs(b - a) > eps:
        y1: float = f(x1)
        y2: float = f(x2)
        if y1 >= y2:
            a = x1
            x1 = x2
            x2 = a + (b - a) / phi
        else:
            b = x2
            x2 = x1
            x1 = b - (b - a) / phi
        k += 1
    xmin: float = (a + b) / 2
    return xmin, k


if __name__ == '__main__':
    a: float = 0.5
    b: float = 1.0
    eps: float = 5e-5
    print(f'xmin = {golden_ratio(a, b, eps)[0]}')
    print(f'k = {golden_ratio(a, b, eps)[1]}')
    x: np.ndarray = np.linspace(1e-3, 3, 200)
    y: np.ndarray = f(x)
    
    plt.title('Golden Ratio')
    plt.xlabel('x')
    plt.ylabel('y')

    plt.plot(x, y, label=f'f(x)\nx_min = {golden_ratio(a, b, eps)[0]}\ny_min = {f(golden_ratio(a, b, eps)[0])}\nk = {golden_ratio(a, b, eps)[1]}')
    
    xmin: float = golden_ratio(a, b, eps)[0]
    ymin: float = f(xmin)

    plt.scatter(xmin, ymin)
    plt.text(xmin, ymin - 0.5, 'min', weight='bold')

    plt.grid()
    plt.legend()
    plt.show()

File: MATLAB_MODULE_1/goldenratio.m
clc
clear variables
close all


fnc = @(x) x^2 + 3 * x * (log(x) - 1);

a = 0.5;
b = 1;
eps = 0.05;

[Fmin, Xmin, k] = golden_ratio(fnc, a, b, eps);
fprintf('Xmin = %e, Fmin = %e, k = %d\n', Xmin, Fmin, k);

X = 0 : 0.01 : 10;
Y = X.^2 + 3 .* X .* (log(X) - 1);
plot(X, Y, 'LineWidth', 1.3, 'DisplayName', 'f(x)');
hold on;
plot(Xmin, Fmin, 'o', 'MarkerSize', 10, 'LineWidth', 1.5, 'DisplayName', 'Fmin');
grid on; grid minor;
xlabel('x'); ylabel('f(x)');
title('Golden Ratio Minimization');
legend show;

function [Fmin, Xmin, k] = golden_ratio(fnc, a, b, eps)
    k = 1;
    phi = (1 + sqrt(5)) / 2;
    x1 = b - (b - a) / phi;
    x2 = a + (b - a) / phi;
    while abs(b - a) > eps
        y1 = fnc(x1);
        y2 = fnc(x2);
        if y1 >= y2
            a = x1;
            x1 = x2;
            x2 = a + (b - a) / phi;
        else
            b = x2;
            x2 = x1;
            x1 = b - (b - a) / phi;
        end
        k = k + 1;
    end
    Xmin = (a + b) / 2;
    Fmin = fnc(Xmin);
end